cfgov/prepaid_agreements/views.py

```diff
@@ -0,0 +1,119 @@
+        ),
+    ).filter(search=search_term)
+    active_filters = {'prepaid_type': [], 'status': [], 'issuer_name': []}
+    for product in products:
+        if product.prepaid_type not in active_filters['prepaid_type']:
+            if product.prepaid_type != '':
+                active_filters['prepaid_type'].append(
+                    product.prepaid_type)
+        if product.status not in active_filters['status']:
+            active_filters['status'].append(product.status)
+        if product.issuer_name not in active_filters['issuer_name']:
+            active_filters['issuer_name'].append(product.issuer_name)
+    return products, active_filters
+
+
+def index(request):
+    filters = dict(request.GET.iterlists())
+    active_filters = {}
+    search_term = None
+    search_field = None
+    products = PrepaidProduct.objects
+    total_count = products.count()
+    valid_filters = [
+        'prepaid_type', 'status', 'issuer_name'
+    ]
+    if filters:
+        filters.pop('page', None)
+        search_term = filters.pop('q', None)
+        search_field = filters.pop('search_field', None)
+
+        if search_term:
+            search_term = search_term[0].strip()
+            if search_term != '':
+                products, active_filters = search_products(
+                    search_term, search_field, products
+                )
+
+        if 'issuer_name' in filters:
+            issuers = Q()
+            for issuer in filters['issuer_name']:
+                issuers |= Q(issuer_name__iexact=issuer)
+            products = products.filter(issuers)
+
+        if 'prepaid_type' in filters:
+            prepaid_types = Q()
+            for prepaid_type in filters['prepaid_type']:
+                prepaid_types |= Q(prepaid_type__iexact=prepaid_type.title())
+            products = products.filter(prepaid_types)
+
+        if 'status' in filters:
+            products = products.filter(status__iexact=filters['status'][0])
+
+    current_count = products.count()
+    paginator = Paginator(products.all(), 20)
+    page_number = validate_page_number(request, paginator)
+    page = paginator.page(page_number)
+
+    if not active_filters:
+        for filter_name in valid_filters:
+            active_filters[filter_name] = PrepaidProduct.objects.order_by(
+                filter_name).values_list(filter_name, flat=True).distinct()
+
+    return render(request, 'prepaid_agreements/index.html', {
+        'current_page': page_number,
+        'results': page,
+        'total_count': total_count,
+        'paginator': paginator,
+        'current_count': current_count,
+        'filters': filters,
+        'query': search_term or '',
+        'active_filters': active_filters,
+        'valid_filters': valid_filters,
+        'search_field': search_field,
+    })
+
+
+def detail(request, product_id):
+    return render(request, 'prepaid_agreements/detail.html', {
+        'product': PrepaidProduct.objects.get(id=product_id)
```

Comment:
This will trigger a `PrepaidProduct.DoesNotExist` exception if the requested ID does not exist in the database, causing a 500 when we probably want a 404 instead.

One approach would be to use the built-in [`get_object_or_404`](https://docs.djangoproject.com/en/1.11/topics/http/shortcuts/#get-object-or-404) like:

```py
get_object_or_404(PrepaidProduct, pk=product_id)
```

This will cause a 404 instead of raising an unhandled exception if the ID is invalid.

Another approach would be to replace this detail view entirely with the built-in Django [`DetailView`](https://docs.djangoproject.com/en/1.11/ref/class-based-views/generic-display/#detailview):

```py
class PrepaidProductDetailView(DetailView):
    model = PrepaidProduct
    template_name = 'prepaid_agreements/detail.html'

# then in urls.py
    ...
    url(r'^detail/(?P<pk>\d+)/$', PrepaidProductDetailView.as_view(), ...),
    ....

Reply:
Thanks! This is fixed in https://github.com/cfpb/cfgov-refresh/commit/e5aea1f71c2ed02ba173aa6a5c7be1c7ed8795c8